### PR TITLE
Add `EmptyEnergyStorage`

### PR DESCRIPTION
**Pull request by Technici4n**
Simple noop implementation to match `EmptyHandler` (items) and `EmptyFluidHandler`. I'd use it in AE2 and Powah.

#### Why the class is not final
It can be useful to write subclasses, here are some examples I found when searching `"extends EmptyFluidHandler`" on GitHub:
- https://github.com/ForestryMC/ForestryMC/blob/30149ccac155aa5b88f7323b39f039a8785a5cba/src/main/java/forestry/core/fluids/FakeTankManager.java#L23
- https://github.com/Kotori316/FluidTank/blob/f1f846f86b41b5e495ecf6cb743b758b278fb518/src/main/scala/com/kotori316/fluidtank/fluids/DebugFluidHandler.scala#L11

### CODE REVIEW

**Review by sciwhiz12 — COMMENTED** [review]
As I've noted in previous discussions on the Discord server, I do not believe it to be wise to allow a singleton-like class such as this to not be made as such: non-extendable through class finality, or a non-public constructor.

One of the main arguments against my suggestion is that mods may wish to extend `EmptyEnergyStorage` for thier own purposes. I disagree, on the principle that those mods are fully able to reimplement what this class does in their own code, without needing reference to this class. Extending this class offers no substantial benefit over being reimplemented by mods.

Furthermore, this class purports to be a specialized implementation of `IEnergyStorage` for the sole purpose of allowing callers of `getCapability(ForgeCapabilities.ENERGY)` to unwrap it into an `IEnergyStorage` variable with the `EmptyEnergyStorage.INSTANCE` field and `Optional#orElse`, instead of needing to use `Optional#ifPresent` or a `null` value and check. With that in mind, I think it ought to be non-extendable with a `private` constructor, to disallow other purposes from extending this class. (Again, mods can make their own implementation as required; it is not in any way difficult to make their own implementation of this class with their own customizations.)

With regards to `EmptyHandler` (for tiems) and `EmptyFluidHandler`, I would suggest (not in this PR) to conform them to the same suggestion as I make now for `EmptyEnergyStorage`.

I would also consider two other alternatives:

- If it is considered to be a base implementation rather than a single-purpose implementation, then a rename to `BaseEnergyStorage` with an `EMPTY` instance stored in a `static final` field; or
- An `EMPTY` instance stored as a field in `IEnergyStorage` itself.

**Comment by Shadows-of-Fire** [comment]
I think this new class should be introduced with the same cadence that `EmptyHandler` and `EmptyFluidHandler` take now. If they should be changed, they should all be changed at once, but they should all be kept in sync.

We may want to defer to core team for this, as this disagreement has been ongoing for a bit.

**Comment by sciwhiz12** [comment]
I hold a bit of hesitance on this PR being accepted on that point, primarily because the fear of breaking mods in the future may cause hesitance on any future PR that would seek to change the three at once. 

I would think that it would be more palatable to do such a change now, rather than do it later and risk the ire of developers who expect something newly introduced to be stable. However, I recognize that we do proper terminal deprecations for that exact reason.

Still, I think that having an `EMPTY` field on the `IEnergyStorage` interface should be considered rather than exposing a new class, if the purpose of this PR is solely to allow `getCapability(ForgeCapabilities.ENERGY).orElse(EMPTY)`.

**Comment by LexManos** [comment]
Making public some discussions from discord, I do not agree with the premise of this class at all. The fact that you want to save yourself a null check at the cost of runtime performance is just not something we should encourage. So the justification for this class is out the window on that one.

However, there is argument to be made about if we should be providing these 'simple' implementations at all. And if we do, what implementations should we do?

My opinion is that we should strip back the exposure level of a lot of these. Limit it down to the Capability Interface, and perhaps a utility class.
For EnergyStorage, if we were to provide any of these functionality. I would propose:
```
  EnergyUtil {
    IEnergyStorage voiding() // Returns an instance that accepts all energy inputs, and disallows all extraction
    IEnergyStorage infinite() // Returns an instance that does not accept any inputs, but allows extraction infinitely.
    IEnergyStorage creative() // Returns an instance that accepts infinite inserts and extractions.
    IEnergyStorage simple(int maxStorage, int maxTransfer) // Simple implementation 
 }
``` 
Limiting the exposed API this way is more inline with a system like [Options.empty()](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--), and [Collections.emptyList()](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#emptyList--). This allows us to hide the implementation details {the classes themselves} behind package private/anonymous implementations. Lowering our exposed API and clearly defining what that API is meant for. {Yes, I know these examples are also because of generics}

IItemHandler implementations would also need reviewing to migrate to this same standard. 
We need a bigger discussion on what these 'simple' implementations of each capability should be.
Each implementation should directly address a CLEARLY defined want from modders, to that, the first step would be getting a list of clearly defined  wants.

**Comment by Technici4n** [comment]
@sciwhiz12 `orElse` was just an exemple, there's other use cases for empty capability implementations. For example: https://github.com/AppliedEnergistics/Applied-Energistics-2/blob/574b4e22b90963075e66e60f184bffe33f1b1b5e/src/main/java/appeng/parts/p2p/FEP2PTunnelPart.java#L45.

I linked some examples that show why we being able to extend this class is useful, be it to also implement custom mod internal interfaces, or for debugging purposes. Of course, modders can reimplement these themselves, but I don't see why Forge should try to limit modders instead of helping them.

@LexManos The existing implementations are all useful in some way. I don't really see the point of having "generic" helper factories, when each capability works in a different way, and has different needs for base implementations. For example creative energy storages work different from creative item/fluid handlers; energy doesn't have slots; fluid handlers don't _really_ have slots either; etc... I'd rather keep the current implementations that all make sense for their cap, and add missing ones.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]
Per discussions with the PR author in Discord[^end], the solution of `EmptyEnergyStorage` as a new class was rejected both by core team decision and later by agreement. The agreed upon resolution is a new field `EMPTY` in `IEnergyStorage` which is an anonymous implementation of the same, having the same behavior as this PR's `EmptyEnergyStorage` -- that is, a no-operation energy storage implementation.

As a more personal side-note, the same change might be considered for the other capabilities in the future, to reduce the effective surface (and maintenance burden) of the Forge API.

[^end]: In the official Forge Discord server, in `#github-discussions`, at [the end of the discussion](https://discord.com/channels/313125603924639766/852298000042164244/1106701439008518206).

**Comment by LexManos** [comment]
I intend to redesign the entire capability api and things that are exposed. Pulling this now as it follows existing paradigm with the fluid system. But expect to see changes in future versions.

**Pull request merged**
